sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/EqualNullSafe.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.expressions.filter;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.expressions.Expression;
+import org.apache.spark.sql.connector.expressions.Literal;
+
+/**
+ * Performs equality comparison, similar to [[EqualTo]]. However, this differs from [[EqualTo]]
+ * in that it returns `true` (rather than NULL) if both inputs are NULL, and `false`
+ * (rather than NULL) if one of the input is NULL and the other is not NULL.
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public final class EqualNullSafe<T> extends Filter {
+  private final Expression column;
+  private final Literal<T> value;
+
+  public EqualNullSafe(Expression column, Literal<T> value) {
+    this.column = column;
+    this.value = value;
+  }
+
+  public Expression column() { return column; }
```

Comment:
> pushdown transforms such as bucket(col, 32) == 4 which could be optimized by data sources

Is it true? AFAIK the current design is users write `WHERE col = 1` and `col` is a partitioned column created with `CREATE TABLE ... PARTITIONED BY year(ts_col)`. The filter pushdown doesn't need to recognize the transform functions, so do end-users when writing queries.

Reply:
My understanding is if a partition column `ts` is of timestamp type with transform `days(ts)`, then yes, data sources like Iceberg can evaluate filters `ts >= a` where `a` is a timestamp, by comparing `a` against transformed partition values. 

However, I think the above only is restricted to date & timestamp types (e.g., `days`, `years`, `hours`, etc)? for the case of `bucket(col, 32) == 4` where `col` is of integer type, data sources would have no way to pushdown the filter without the transform information, is that correct?

Reply:
`bucket` transform is for the bucketed table feature: `CREATE TABLE ... CLUSTERED BY col INTO 8 BUCKETS`. Then users can write queries like `WHERE col = 1`, and the v2 table knows the table is partitioned by `bucket(col, 8)`, so that it can optimize the predicate `col = 1`.

Reply:
In general, a partitioned table is like we create an index for a table with one or more columns, and users can run queries faster if they use predicates with the column/columns that are indexed.

Reply:
Thanks, what you said makes sense. But what if we have `bucket(col, 32) = 10` where `col` is of string type?

Reply:
right now do users can write `WHERE bucket(col, 32) = 10` in Spark?

Reply:
Currently no, but we think it can be an optimization in future, by allowing V2 functions to appear in the filters. As mentioned earlier, another use case is function pushdown. For example, I've also being told that in Druid, some functions related to timestamp can be efficiently evaluated on the data source level.

IMO it's better to leave more room in the API design so it can be extended for more use cases such as the above.

Reply:
OK then the name `column` doesn't make sense. How about `expr`?

Reply:
changed to `expr`

Reply:
I created https://issues.apache.org/jira/browse/SPARK-36695 to track this. I still think it is better not to make this API too restrictive. Perhaps we can think on it a bit more over the 3.3.0 time frame.